samza-test/src/test/java/org/apache/samza/test/startpoint/StartpointTestHarness.java

```diff
@@ -0,0 +1,468 @@
+import org.apache.samza.startpoint.StartpointManager;
+import org.apache.samza.startpoint.StartpointOldest;
+import org.apache.samza.startpoint.StartpointSpecific;
+import org.apache.samza.startpoint.StartpointTimestamp;
+import org.apache.samza.startpoint.StartpointUpcoming;
+import org.apache.samza.system.IncomingMessageEnvelope;
+import org.apache.samza.system.SystemAdmin;
+import org.apache.samza.system.SystemAdmins;
+import org.apache.samza.system.SystemStream;
+import org.apache.samza.system.SystemStreamPartition;
+import org.apache.samza.task.TaskCallback;
+import org.apache.samza.test.StandaloneTestUtils;
+import org.apache.samza.test.harness.IntegrationTestHarness;
+import org.apache.samza.test.processor.SharedContextFactories;
+import org.apache.samza.test.processor.TestTaskApplication;
+import org.apache.samza.test.util.TestKafkaEvent;
+import org.apache.samza.util.CoordinatorStreamUtil;
+import org.apache.samza.util.NoOpMetricsRegistry;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.rules.Timeout;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotEquals;
+import static org.junit.Assert.assertTrue;
+
+
+public class StartpointTestHarness extends IntegrationTestHarness {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(StartpointTestHarness.class);
+
+  private static final int NUM_KAFKA_EVENTS = 300;
+  private static final int ZK_TEST_PARTITION_COUNT = 5;
+  private static final String TEST_SYSTEM = "TestSystemName";
+  private static final String TEST_SSP_GROUPER_FACTORY = "org.apache.samza.container.grouper.stream.GroupByPartitionFactory";
+  private static final String TEST_TASK_GROUPER_FACTORY = "org.apache.samza.container.grouper.task.GroupByContainerIdsFactory";
+  private static final String TEST_JOB_COORDINATOR_FACTORY = "org.apache.samza.zk.ZkJobCoordinatorFactory";
+  private static final String TEST_SYSTEM_FACTORY = "org.apache.samza.system.kafka.KafkaSystemFactory";
+  private static final String TASK_SHUTDOWN_MS = "10000";
+  private static final String JOB_DEBOUNCE_TIME_MS = "10000";
+  private static final String BARRIER_TIMEOUT_MS = "10000";
+  private static final String[] PROCESSOR_IDS = new String[] {"0000000000", "0000000001", "0000000002", "0000000003"};
+
+  private String inputKafkaTopic1;
+  private String inputKafkaTopic2;
+  private String inputKafkaTopic3;
+  private String inputKafkaTopic4;
+  private String outputKafkaTopic;
+  private ApplicationConfig applicationConfig1;
+  private ApplicationConfig applicationConfig2;
+  private ApplicationConfig applicationConfig3;
+  private ApplicationConfig applicationConfig4;
+  private String testStreamAppName;
+  private String testStreamAppId;
+
+  @Rule
+  public Timeout testTimeOutInMillis = new Timeout(150000, TimeUnit.MILLISECONDS);
```

Comment:
Do we need such a long timeout? Is this inferred through some calculation from the debounce, barrier timeout etc? If so, can you add a doc?

Reply:
Will figure out the right timeout and add doc. Good suggestion.